numpy/lib/index_tricks.py

```diff
@@ -596,6 +597,419 @@ def __iter__(self):
+        parameter along each axis.
+    step: tuple of ints (or single int)
+        If provided, must be the same length as stop. Specifies the increment
+        along each axis. If step is not provided, it defaults to a tuple of
+        ones.
+
+    See Also
+    --------
+    flatiter, ndenumerate, ndindex
+
+    Notes
+    -----
+    .. versionadded:: 1.16.0
+
+    In Python 2.7, this object is based ``N = len(stop)`` range objects,
+    ``N`` lists, as opposed to ``N`` xrange objects to allow for slicing.
+
+    In addition to this, because neither `range` nor `xrange` are hashable
+    in Python 2.7, the ndrange object is not hashable either in Python 2.7.
+
+    Examples
+    --------
+    This is likely the most natural way to iterate through the sequence of
+    ndrange object. Here, the first parameter is the shape of the array
+    ``shape`` attribute of an array and we slice through the ``ndrange``
+    object.
+
+    >>> a = np.zeros((4, 4), dtype='object')
+    >>> # [...] # operate on ``a``
+    >>> for i in np.ndrange(a.shape)[::2, 1:-1]:
+    ...     print(i)
+    (0, 1)
+    (0, 2)
+    (2, 1)
+    (2, 2)
+
+    >>> list(np.ndrange((2, 3)))
+    [(0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2)]
+
+    >>> list(np.ndrange((1, 2), (4, 4)))
+    [(1, 2), (1, 3), (2, 2), (2, 3), (3, 2), (3, 3)]
+
+    Reversed iteration is also possible
+
+    >>> for i in reversed(np.ndrange(3, 2)):
+    ...     print(i)
+    (2, 1)
+    (2, 0)
+    (1, 1)
+    (1, 0)
+    (0, 1)
+    (0, 0)
+
+    Keep in mind that ``ndrange`` is not an iterator. To obtain an iterator,
+    call ``iter`` first.
+
+    >>> ndrange_iter = iter(np.ndrange((2, 2)))
+    >>> next(ndrange_iter)
+    >>> for i in ndrange_iter:
+    ...     print(i)
+    (0, 1)
+    (1, 0)
+    (1, 1)
+
+    Equality comparison is also possible. Notice that
+
+    >>> np.ndrange((0, 0), (3, 3), (2, 2)) == np.ndrange((0, 0), (4, 4), (2, 2))
+    True
+
+    returns ``True``, since the sequence generated by both those pairs of
+    ``ndrange`` objects would be the same.
+
+    Testing for containment is also possible
+
+    >>> (1, 3) in np.ndrange((1, 2), (4, 4))
+    True
+
+    >>> (0, 3) in np.ndrange((1, 2), (4, 4))
+    False
+
+
+    While native support for Fortran iteration ordering isn't supported, it can
+    be acheived readily through:
+
+    >>> for i in np.ndrange((3, 2)[::-1]):
+    ...     i = i[::-1]
+    ...     print(i)
+    (0, 0)
+    (1, 0)
+    (2, 0)
+    (0, 1)
+    (1, 1)
+    (2, 1)
+
+    or
+
+    >>> for i_f in (i_c[::-1] for i_c in np.ndrange((3, 2)[::-1])):
+    ...     print(i_f)
+    (0, 0)
+    (1, 0)
+    (2, 0)
+    (0, 1)
+    (1, 1)
+    (2, 1)
+
+    The conceptual ``r`` can be obtained by making a call to ``np.asarray``.
+    We strongly against this method of array creation since it is slow.
+
+    >>> np.asarray(np.ndrange((4, 4))[::2, 1:-1])
+    array([[(0, 1), (0, 2)],
+           [(2, 1), (2, 2)]], dtype=object)
```

Comment:
Honestly, I would prefer raising an error. That would allow us to kick this down the road until somebody finds a meaningful usecase that isn't covered by ogrid or meshgrid.